### DISCUSSION TITLE
x-with?

### DISCUSSION THREAD

**Discussion by Lomacar** [alpinejs/alpine#2119] [1. Help]
Is there some way in Alpine to something like `with` in Javascript? That is, focus on one branch of whatever is contained in x-data within a certain html element? I thought something like this would work (nested within a larger x-data context), but apparently not.

    <div x-data="data.user[0].name">
        <p x-text="firstname"></p>
        <p x-text="lastname"></p>
    </div>

I also tried using `x-data="$root.data..."`.

Oh, I should mention that `data` is some AJAX data and it is loading properly. If I just say `x-text="data.user[0].name.firstname"` it works.

**Comment by SimoTod** [comment]
Nothing out of the box but you can create your own directive: https://codepen.io/SimoTod/pen/bGRxQZQ

It would be even simpler if Alpine was passing the `addScopeToNode` utility to custom directives. 

@calebporzio do you think it would be possible to add `addScopeToNode: addScopeToNode.bind(addScopeToNode, el)` to https://github.com/alpinejs/alpine/blob/v3.4.1/packages/alpinejs/src/directives.js#L81? It's mostly to avoid using an internal property like _x_dataStack

**Reply by SimoTod** [reply]
I'm happy to prepare a PR and update the docs if it' okay

**Comment by calebporzio** [comment]
Before we PR this, what you posted should actually work. Here is a working snippet I just tested:

```html
<div x-data="{ data: { foo: 'bar' } }">
    <div x-data="data">
        <span x-text="foo"></span>
    </div>
</div>
```

**Reply by SimoTod** [reply]
Sweet, I just assumed it wasn't working. 😅 Im that case, no need for addScopeToNode for now.

**Reply by Lomacar** [reply]
Oh, well that's great, but why doesn't it work for me with this AJAX data? https://codepen.io/lomacar/pen/ZEymvRb

Am I doing something daft?

**Reply by Tim-Wils** [reply]
@Lomacar Hmmm weird, didn't expect that, but it looks like async data has to be re-rendered after the fetch. This is done by wrapping your divs in a template-tag with a x-if on the books key, so the expressions are parsed afterwards. My fiddle: https://jsfiddle.net/4n9vqg63/

**Reply by Lomacar** [reply]
OK, I don't really understand why it works that way, but thanks. It seems that Alpine doesn't work so great with async data.

**Reply by Lomacar** [reply]
Is there a way from the init() in my grab() function to force the re-render?

**Reply by Tim-Wils** [reply]
I agree this feels weird, but maybe this is a bug and @calebporzio can shine his light on this one?

**Reply by SimoTod** [reply]
I don't think it's weird. You are initialising the parent component that, to begin with, doesn't have books since it's set by init but, being async, it won't wait until it has finished before processing the rest. 
Then the first component tries to access something that doesn't even exist so it can't really start up nor set any reactivity.
What were you expecting it to show before the ajax call was completed for instance?

**Reply by Tim-Wils** [reply]
Sorry, you're right! :) Guess I was thinking in a wrong way here

**Reply by Lomacar** [reply]
@SimoTod I would expect that when the inner x-data gets its data then its children would automatically update. Before that I would expect it to show whatever dummy content might be in the children.

**Reply by SimoTod** [reply]
Unfortunately, if you refer to a property and that property doesn't exists, that component errors and then it never updates afterwards. It would error in js as well. It may work if you use the conditional operator `?.` but it's not supported by all browsers. It will never keep the dummy content anyway: X-text will always set the content (except when it crashes, of course)

**Reply by Lomacar** [reply]
OK, if it was just about an error then I would expect/hope for this to still work: `x-data="books ? books.items[0].volumeInfo : {books:{}, title:'title'}"` It does of course load `x-text="title"` fine, but it never updates when `books` is populated because I guess it is never re-evaluated. Regardless of errors or not it would be nice to have a way to re-evaluate elements that are hooked into Alpine.

**Reply by SimoTod** [reply]
No x-data only evaluates once. You need the conditional in x-text. Something like https://codepen.io/SimoTod/pen/gORyEWZ

In general, you should never use variables that are not declared first.
